source/opt/aggressive_dead_code_elim_pass.h

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2017 The Khronos Group Inc.
+// Copyright (c) 2017 Valve Corporation
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_AGGRESSIVE_DCE_PASS_H_
+#define LIBSPIRV_OPT_AGGRESSIVE_DCE_PASS_H_
+
+#include <algorithm>
+#include <map>
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include "basic_block.h"
+#include "def_use_manager.h"
+#include "module.h"
+#include "pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class AggressiveDCEPass : public Pass {
+
+  using cbb_ptr = const ir::BasicBlock*;
+
+ public:
+   using GetBlocksFunction =
+     std::function<std::vector<ir::BasicBlock*>*(const ir::BasicBlock*)>;
+
+  AggressiveDCEPass();
+  const char* name() const override { return "aggressive-dce"; }
+  Status Process(ir::Module*) override;
+
+ private:
+  // Returns true if |opcode| is a non-ptr access chain op
+  bool IsNonPtrAccessChain(const SpvOp opcode) const;
+
+  // Given a load or store |ip|, return the pointer instruction.
+  // Also return the base variable's id in |varId|.
+  ir::Instruction* GetPtr(ir::Instruction* ip, uint32_t* varId);
+
+  // Add all store instruction which use |ptrId|, directly or indirectly,
+  // to the live instruction worklist.
+  void AddStores(uint32_t ptrId);
+
+  // Return true if variable with |varId| is function scope
+  bool IsLocalVar(uint32_t varId);
+
+  // For function |func|, mark all Stores to non-function-scope variables
+  // and block terminating instructions as live. Recursively mark the values
+  // they use. When complete, delete any non-live instructions. Return true
```

Comment:
This is too aggressive, e.g. potentially sweeping away:
- synchronization instructions like OpMemoryBarrier and OpControlBarrier
- extended instructions.   Some, like Frexp, can modify storage. 
- unknown extended instructions.
- collective functions like SubgroupBallotKHR
- atomic  instructions

Perhaps, ugly though it may be, you should use a whitelist of known "safe" instructions.  An instruction is safe if it is a combinator: its effect is purely determined by its arguments, and has no side effects.   It seems the kind of list that should be centralized. (In the long run, maybe it's a fundamental property that ought to be captured in the SPIR-V grammar files.)